// @custom-js/n8n-nodes-pdf-toolkit/dist/nodes/SSLChecker/SSLChecker.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSLChecker = void 0;
class SSLChecker {
    constructor() {
        this.description = {
            displayName: "SSL Checker (SSL Certificate) (CustomJS)",
            name: "sslChecker",
            icon: "file:customJs.svg",
            group: ["transform"],
            version: 1,
            description: "Gives you information about the expiration date of SSL certificates.",
            defaults: {
                name: "SSL Checker",
            },
            inputs: ["main"],
            outputs: ["main"],
            credentials: [
                {
                    name: "customJsApi",
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: "Domain",
                    name: "domain",
                    type: "string",
                    default: "",
                    description: "The domain to check (e.g. example.com)",
                    required: true,
                    modes: [
                        {
                            displayName: "By Domain",
                            name: "domain",
                            type: "string",
                            placeholder: "example.com",
                        },
                    ],
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        for (let i = 0; i < items.length; i++) {
            const credentials = await this.getCredentials("customJsApi");
            const domain = this.getNodeParameter("domain", i);
            const options = {
                url: `https://e.customjs.io/__js1-${credentials.apiKey}`,
                method: "POST",
                headers: {
                    "customjs-origin": "n8n/sslchecker",
                    "x-api-key": credentials.apiKey,
                },
                body: {
                    input: domain,
                    code: "const checkCertExpiration = require('check-cert-expiration'); return checkCertExpiration(input);",
                },
                json: true,
            };
            const response = await this.helpers.request(options);
            returnData.push({
                json: {
                    output: response,
                },
            });
        }
        return [returnData];
    }
}
exports.SSLChecker = SSLChecker;
